src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,219 @@
+        (index + query_index) == uri.query_index) {
+        // We had some cases where this was expected behavior
+        // We identify them by checking whether the next character is a '&'
+        auto after_param = uri.raw.substr(index + param.length());
+        if (!after_param.empty() && after_param[0] == '&') {
+            return false;
+        }
+
+        // We can also find them if it is followed by a parameter (abc=42)
+        for (std::size_t i = 0; i < after_param.size(); ++i) {
+            const auto c = after_param[i];
+            if (c == '=' && i > 0) {
+                return false;
+            }
+
+            if (!ddwaf::isalnum(c) && c != '_') {
+                break;
+            }
+        }
+
+        return true;
+    }
+
+    // Check if the parameter interfered with what comes after the ?
+
+    // The parameter stop before the ? or start after #, so it can't be interfering with what come
+    // after. We can stop processing this heuristic
+    const auto param_end = index + param.size() - 1;
+    if (uri.query_index > param_end || (uri.fragment_index != npos && uri.fragment_index < index)) {
+        return false;
+    }
+
+    // We compute the substring of the parameter that come after the ?
+
+    auto query_param = uri.query_index <= index ? param : param.substr(uri.query_index - index);
+    return query_param.find('&') != npos;
+}
+
+ssrf_result ssrf_impl(const uri_decomposed &uri, const ddwaf_object &params,
+    const exclusion::object_set_ref &objects_excluded, const object_limits &limits,
+    const std::unique_ptr<matcher::ip_match> &dangerous_ip_matcher,
+    const std::unordered_set<std::string_view> &authorised_scheme_set, ddwaf::timer &deadline)
+{
+    static constexpr std::size_t min_str_len = 4;
+
+    std::string_view dangerous_domain = {};
+    for (const auto domain : dangerous_domains) {
+        if (uri.authority.raw.ends_with(domain)) {
+            dangerous_domain = domain;
+        }
+    }
+
+    std::optional<ssrf_result> parameter_injection;
+
+    object::kv_iterator it(&params, {}, objects_excluded, limits);
+    for (; it; ++it) {
+        if (deadline.expired()) {
+            throw ddwaf::timeout_exception();
+        }
+
+        const ddwaf_object &param = *(*it);
+        if (param.type != DDWAF_OBJ_STRING || param.nbEntries < min_str_len) {
+            continue;
+        }
+
+        std::string_view value{param.stringValue, static_cast<std::size_t>(param.nbEntries)};
+        auto index = uri.raw.find(value);
+        if (index == std::string_view::npos) {
+            // Seemingly no injection
+            continue;
+        }
+
+        // Verify if the injected value intereferes with the authority
+        if (index < uri.scheme_and_authority.size() &&
+            (index + value.size()) > (uri.scheme.size() + 3)) {
+            // Verify if the host was modified by the injected value
+            if ((index + value.size() - 1) > uri.scheme_and_authority.size()) {
```

Comment:
This slightly odd condition is meant to remove certain false positives caused by injections such as `http://domain.com/` where the injection is `domain.com/`.

I can only assume this was introduced as a result of some empirical false positives rather than assumed ones.